src/sss_client/pam_sss_gss.c

```diff
@@ -0,0 +1,477 @@
+                                     gss_name, GSS_C_NO_OID, flags, 0, NULL,
+                                     &input, NULL, &output,
+                                     &ret_flags, NULL);
+
+        free(input.value);
+        memset(&input, 0, sizeof(gss_buffer_desc));
+
+        if (major == GSS_S_CONTINUE_NEEDED || output.length > 0) {
+            ret = sssd_establish_sec_ctx_send(pamh, pam_service, domain,
+                                              output.value, output.length,
+                                              &input.value, &input.length);
+            if (ret != EOK) {
+                goto done;
+            }
+        }
+
+        gss_release_buffer(NULL, &output);
+
+        if (GSS_ERROR(major)) {
+            ERROR(pamh, "Unable to establish GSS context [maj:0x%x, min:0x%x]",
+                  major, minor);
+            gssapi_log_status(pamh, GSS_C_GSS_CODE, major);
+            gssapi_log_status(pamh, GSS_C_MECH_CODE, minor);
+            ret = EIO;
+            goto done;
+        }
+
+        switch (major)
+        {
+        case GSS_S_COMPLETE:
+            established = true;
+            break;
+        case GSS_S_CONTINUE_NEEDED:
+            break;
+        default:
+            ERROR(pamh, "Context is not established but major has "
+                  "unexpected value: %x\n", major);
+            ret = EIO;
+            goto done;
+        }
+    }
+
+    if ((ret_flags & flags) != flags) {
+        ERROR(pamh, "Negotiated context does not support requested flags\n");
+        ret = EIO;
+        goto done;
+    }
+
+    ret = EOK;
+
+done:
+    gss_delete_sec_context(&minor, &ctx, NULL);
+    gss_release_name(&minor, &gss_name);
+
+    return ret;
+}
+
+static int errno_to_pam(pam_handle_t *pamh, errno_t ret)
+{
+    switch (ret) {
+        case EOK:
+            TRACE(pamh, "Authentication successful");
+            return PAM_SUCCESS;
+        case ENOENT:
+            TRACE(pamh, "User not found");
+            return PAM_USER_UNKNOWN;
+        case ENOTSUP:
+            TRACE(pamh, "GSSAPI authentication is not enabled "
+                  "for given user and service");
+            return PAM_USER_UNKNOWN;
+        case ESSS_NO_SOCKET:
+            TRACE(pamh, "SSSD socket does not exist");
+            return PAM_AUTHINFO_UNAVAIL;
+        case EPERM:
+            TRACE(pamh, "Authentication failed");
+            return PAM_AUTH_ERR;
+        default:
+            TRACE(pamh, "System error [%d]: %s",
+                     ret, strerror(ret));
+            return PAM_SYSTEM_ERR;
+    }
+}
+
+int pam_sm_authenticate(pam_handle_t *pamh,
+                        int flags,
+                        int argc,
+                        const char **argv)
+{
+    const char *pam_service;
+    const char *pam_user;
+    char *domain = NULL;
+    char *target = NULL;
+    uid_t uid;
+    uid_t euid;
+    errno_t ret;
+    int i;
```

Comment:
Counter variable isn't needed outside scope - this is a C99 codebase, so initialize in the loop: `for (int i = 0; i < argc; i++) {`.